daemons/based/pacemaker-based.c

```diff
@@ -52,141 +55,149 @@ static bool startCib(const char *filename);
+               GError **error)
+{
+    int rc = 0;
+    struct passwd *pwentry = NULL;
+
+    stand_alone = TRUE;
+    preserve_status = TRUE;
+    cib_writes_enabled = FALSE;
+
+    pwentry = getpwnam(CRM_DAEMON_USER);
+    CRM_CHECK(pwentry != NULL,
+              exit_code = CRM_EX_FATAL;
+              g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                          "Invalid uid (%s) specified", CRM_DAEMON_USER);
+              return FALSE);
+
+    rc = setgid(pwentry->pw_gid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not set group to %d", pwentry->pw_gid);
+        return FALSE;
+    }
+
+    rc = initgroups(CRM_DAEMON_USER, pwentry->pw_gid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not setup groups for user %d", pwentry->pw_uid);
+        return FALSE;
+    }
+
+    rc = setuid(pwentry->pw_uid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not set user to %d", pwentry->pw_uid);
+        return FALSE;
+    }
+    return TRUE;
+}
+
+static GOptionEntry entries[] = {
+    { "stand-alone", 's', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK,
+      stand_alone_cb, "(Advanced use only) Run in stand-alone mode", NULL },
+
+    { "disk-writes", 'w', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE,
+      &cib_writes_enabled,
+      "(Advanced use only) Enable disk writes (enabled by default unless in "
+      "stand-alone mode)", NULL },
+
+    { "cib-root", 'r', G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING, &cib_root,
+      "(Advanced use only) Directory where the CIB XML file should be located "
+      "(default: " CRM_CONFIG_DIR ")", NULL },
+
+    { NULL }
 };
 
+static GOptionContext *
+build_arg_context(pcmk__common_args_t *args)
+{
+    GOptionContext *context = NULL;
+
+    context = pcmk__build_arg_context(args, NULL, NULL, "[metadata]");
+    pcmk__add_main_args(context, entries);
+    return context;
+}
+
 int
 main(int argc, char **argv)
 {
-    int flag;
-    int rc = 0;
-    int index = 0;
-    int argerr = 0;
-    struct passwd *pwentry = NULL;
     crm_ipc_t *old_instance = NULL;
 
+    GError *error = NULL;
+
+    pcmk__common_args_t *args = pcmk__new_common_args(SUMMARY);
+    gchar **processed_args = pcmk__cmdline_preproc(argv, NULL);
+    GOptionContext *context = build_arg_context(args);
+
     crm_log_preinit(NULL, argc, argv);
-    pcmk__set_cli_options(NULL, "[options]", long_options,
-                          "daemon for managing the configuration "
-                          "of a Pacemaker cluster");
 
-    mainloop_add_signal(SIGTERM, cib_shutdown);
-    mainloop_add_signal(SIGPIPE, cib_enable_writes);
+    if (!g_option_context_parse_strv(context, &processed_args, &error)) {
+        if (exit_code == CRM_EX_OK) {
```

Comment:
This lets us exit with `CRM_EX_FATAL` (preserving existing behavior) if we hit a user/group error in the callback. That seems more correct than a usage error anyway -- but we could just return FALSE from the callback and throw a usage error here.